### PR TITLE
storage: Create stratis filesystem snapshot without requiring mount_point

### LINKED ISSUE #18718: Allow creating stratis snapshot without selecting mouting point
When I want to create shapshot of stratis filesystem I can select from two actions - "Create snapshot and mount" or "Create snapshot only". When I select the latter I still have to select mount point. This behavior is somewhat confusing.

justsnap

### PR DESCRIPTION
**Pull request by MahmoudHamdy02**
Fixes #18718.

I passed the `variant` from the footer buttons to the `validate` function for the mount point, so it checks the type before validating.
Let me know if this feels like a long workaround or if there is a better approach!

### CODE REVIEW

**Comment by martinpitt** [comment]
Thanks @MahmoudHamdy02 ! I leave the review of the implementation to Marius. This should get covered in test/verify/check-storage-stratis . That already creates file systems, mounts and unmounts them, so hopefully it has enough examples (feel free to watch them with `TEST_SHOW_BROWSER=1` to get a feel for it). Please let me know if you have trouble with that and want/need help, we can do that in our Matrix chat.

**Comment by MahmoudHamdy02** [comment]
@martinpitt I've extended the test to include unmounted filesystems and removed the pixel test since it won't match due to the empty mount point. Should a new reference be made, or is it already covered by the test before it?

File: pkg/storaged/dialog.jsx [hunk 1]
```diff
@@ -379,6 +380,7 @@ export const dialog_open = (def) => {
     };
 
     function run_action(progress_callback, variant) {
+        clicked_variant = variant;
```
Comment (mvollmer) [inline pkg/storaged/dialog.jsx:383]: I think we can just pass `variant` to `validate` as an argument. No need to hide it in a global variable.

**Comment by mvollmer** [comment]
Thanks for working on this!

I think we should implement this behavior also for the regular Format dialog (in `format-dialog.jsx`), not just for Stratis.

File: pkg/storaged/stratis-details.jsx [hunk 2]
```diff
@@ -520,7 +520,10 @@ export const StratisPoolDetails = ({ client, pool }) => {
                               }),
                     TextInput("mount_point", _("Mount point"),
                               {
-                                  validate: val => is_valid_mount_point(client, null, val)
+                                  validate: (val, values, variant) => {
```
Comment (mvollmer) [inline pkg/storaged/stratis-details.jsx:523]: This only allows the empty string as the mountpoint, but the fstab entry is still written later in `set_mount_options`. That function turns a empty string into "/", actually. :)
So we also need to change `set_mount_options` to not do anything when the mount point is left empty.

**Review by mvollmer — CHANGES_REQUESTED** [review]

File: test/verify/check-storage-stratis [hunk 3]
```diff
@@ -161,6 +161,20 @@ class TestStorageStratis(StorageCase):
         self.confirm()
         b.wait_not_in_text("#detail-content", "fsys2-copy")
 
+        # Make an unmounted copy of the filesystem
```
Comment (mvollmer) [inline test/verify/check-storage-stratis:164]: Let's only have one commit with changes to both the code and the tests.

File: pkg/storaged/stratis-details.jsx (hunk 2 above)
Comment (MahmoudHamdy02) [reply pkg/storaged/stratis-details.jsx:523]: How about not calling `set_mount_options` from the start? I think putting the rest of the function (starting from line 309) into a big `if` would mess up the return values.

File: pkg/storaged/dialog.jsx (hunk 1 above)
Comment (MahmoudHamdy02) [reply pkg/storaged/dialog.jsx:383]: Right, updated :grin:

File: test/verify/check-storage-stratis (hunk 3 above)
Comment (MahmoudHamdy02) [reply test/verify/check-storage-stratis:164]: Alright, I'll squash when the rest of the changes are done :+1:

**Comment by MahmoudHamdy02** [comment]
@mvollmer I've pushed the requested changes and did the same fix for creating and formatting unmounted filesystems, I only had one question which I put in the [comment above](https://github.com/cockpit-project/cockpit/pull/18742#discussion_r1184395378).

Should I extend the test to include the unmounted 'create' and 'format' as well?

File: pkg/storaged/dialog.jsx (hunk 1 above)
Comment (mvollmer) [reply pkg/storaged/dialog.jsx]: Perfect!

**Comment by mvollmer** [comment]
> Should I extend the test to include the unmounted 'create' and 'format' as well?

Yes, please!

**Review by mvollmer — CHANGES_REQUESTED** [review]
(see unresolved conversations.)

**Comment by mvollmer** [comment]
Let's run the fedora-37/devel tests, they will give us the code coverage.

File: pkg/storaged/format-dialog.jsx [hunk 4]
```diff
@@ -484,17 +487,19 @@ function format_dialog_internal(client, path, start, size, enable_dos_extended,
...
-                        dir: { t: 'ay', v: utils.encode_filename(mount_point) },
-                        type: { t: 'ay', v: utils.encode_filename("auto") },
-                        opts: { t: 'ay', v: utils.encode_filename(mount_options.join(",") || "defaults") },
-                        freq: { t: 'i', v: 0 },
-                        passno: { t: 'i', v: 0 },
-                        "track-parents": { t: 'b', v: true }
-                    }]);
+                    if (mount_point != "") {
+                        if (mount_point[0] != "/")
+                            mount_point = "/" + mount_point;
```
Comment (cockpituous) [inline pkg/storaged/format-dialog.jsx:492]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18742-20230505-075100-4fa0bcd7-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/stratis-details.jsx (hunk 2 above)
Comment (MahmoudHamdy02) [reply pkg/storaged/stratis-details.jsx:528]: I edited `set_mount_options` as it would be a more general solution

**Comment by MahmoudHamdy02** [comment]
@mvollmer Should be all good now, let me know if this needs any more changes!

File: pkg/storaged/format-dialog.jsx (hunk 4 above)
Comment (mvollmer) [reply pkg/storaged/format-dialog.jsx:492]: This is fine.

**Review by mvollmer — APPROVED** [review]
Excellent!

**Review by mvollmer — CHANGES_REQUESTED** [review]
Code is excellent, but the commit message needs some work. :-)  I missed that earlier, sorry.

**Comment by mvollmer** [comment]
The commit message shouldn't contain the headlines of the squashed commits. We want to have one 'perfect' commit on the main branch that doesn't mention any of the history of how we came up with it.

So something like

```
storage: Don't require a mount point when creating filesystems

When a user explicitly does not want to mount a newly created filesystem, it is surprising that Cockpit forces them to specify a mount point anyway. Let's allow a empty mount point in this case, and not write a fstab entry at all.
```

File: pkg/storaged/format-dialog.jsx (hunk 4 above)
Comment (cockpituous) [inline pkg/storaged/format-dialog.jsx:492]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18742-20230508-073603-1749d05b-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by MahmoudHamdy02** [comment]
@mvollmer Updated :)

File: pkg/storaged/format-dialog.jsx (hunk 4 above)
Comment (cockpituous) [inline pkg/storaged/format-dialog.jsx:492]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18742-20230508-115703-7c10f358-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by mvollmer — APPROVED** [review]
